# EDA

#### Import libraries

In [1]:
%run import_libraries.py
hlp = Helper()

#### Load data

In [2]:
raw_df = pd.read_csv('data/downloaded/train_data.csv')
raw_df.head()

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  ... payment_type  \
0  34.938093  -9.856322                  none            0  ...     annually   
1  34.698766  -2.147466              Zahanati            0  ...    never pay   
2  37.460664  -3.821329           Kwa Mahundi            0  ...   per bucket   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0  ...    never pay   
4  31.130847  -1.825359               Shuleni            0  ...    never pay   

  water_quality quality_group      quantity  quantity_group  \
0          soft          good        enough          enough   
1          soft          good  insufficient    insufficient   
2          soft          good        enough          enough   
3          soft          good           dry             dry   
4          soft          good      seasonal        seasonal   

                 source           source_type  source_class  \
0                spring                spring   groundwater   
1  rainwater harvesting  rainwater harvesting       surface   
2                   dam                   dam       surface   
3           machine dbh              borehole   groundwater   
4  rainwater harvesting  rainwater harvesting       surface   

               waterpoint_type waterpoint_type_group  
0           communal standpipe    communal standpipe  
1           communal standpipe    communal standpipe  
2  communal standpipe multiple    communal standpipe  
3  communal standpipe multiple    communal standpipe  
4           communal standpipe    communal standpipe  

[5 rows x 40 columns]

In [3]:
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 40 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   amount_tsh             59400 non-null  float64
 2   date_recorded          59400 non-null  object 
 3   funder                 55765 non-null  object 
 4   gps_height             59400 non-null  int64  
 5   installer              55745 non-null  object 
 6   longitude              59400 non-null  float64
 7   latitude               59400 non-null  float64
 8   wpt_name               59400 non-null  object 
 9   num_private            59400 non-null  int64  
 10  basin                  59400 non-null  object 
 11  subvillage             59029 non-null  object 
 12  region                 59400 non-null  object 
 13  region_code            59400 non-null  int64  
 14  district_code          59400 non-null  int64  
 15  lg

#### Check and fill na values in categorical features

In [4]:
raw_df_na_cols = raw_df.columns[raw_df.isna().any()].tolist()
print('{0:20}{1:10}{2}\n'.format('Column name', 'Type', 'Filled NA values count'))

for col in raw_df_na_cols:
    raw_df[col] = raw_df[col].fillna('not_filled')
    print('{0:20}{1:10}{2}'.format(
        col,
        str(raw_df[col].dtype),
        len(raw_df[raw_df[col] == "not_filled"])
    ))

Column name         Type      Filled NA values count

funder              object    3635
installer           object    3655
subvillage          object    371
public_meeting      object    3334
scheme_management   object    3877
scheme_name         object    28166
permit              object    3056


#### Explore data

In [5]:
for feature in raw_df:
    print(f'{feature}: {raw_df[feature].nunique()} \n{raw_df[feature].unique()} \n\n')

id: 59400 
[69572  8776 34310 ... 37057 31282 26348] 


amount_tsh: 98 
[6.00e+03 0.00e+00 2.50e+01 2.00e+01 2.00e+02 5.00e+02 5.00e+01 4.00e+03
 1.50e+03 6.00e+00 2.50e+02 1.00e+01 1.00e+03 1.00e+02 3.00e+01 2.00e+03
 4.00e+02 1.20e+03 4.00e+01 3.00e+02 2.50e+04 7.50e+02 5.00e+03 6.00e+02
 7.20e+03 2.40e+03 5.00e+00 3.60e+03 4.50e+02 4.00e+04 1.20e+04 3.00e+03
 7.00e+00 2.00e+04 2.80e+03 2.20e+03 7.00e+01 5.50e+03 1.00e+04 2.50e+03
 6.50e+03 5.50e+02 3.30e+01 8.00e+03 4.70e+03 7.00e+03 1.40e+04 1.30e+03
 1.00e+05 7.00e+02 1.00e+00 6.00e+01 3.50e+02 2.00e-01 3.50e+01 3.06e+02
 8.50e+03 1.17e+05 3.50e+03 5.20e+02 1.50e+01 6.30e+03 9.00e+03 1.50e+02
 1.20e+05 1.38e+05 3.50e+05 4.50e+03 1.30e+04 4.50e+04 2.00e+00 1.50e+04
 1.10e+04 5.00e+04 7.50e+03 1.63e+04 8.00e+02 1.60e+04 3.00e+04 5.30e+01
 5.40e+03 7.00e+04 2.50e+05 2.00e+05 2.60e+04 1.80e+04 2.60e+01 5.90e+02
 9.00e+02 9.00e+00 1.40e+03 1.70e+05 2.20e+02 3.80e+04 2.50e-01 1.20e+01
 6.00e+04 5.90e+01] 


date_recorded: 356 
['2011-03

extraction_type_class: 7 
['gravity' 'submersible' 'handpump' 'other' 'motorpump' 'wind-powered'
 'rope pump'] 


management: 12 
['vwc' 'wug' 'other' 'private operator' 'water board' 'wua' 'company'
 'water authority' 'parastatal' 'unknown' 'other - school' 'trust'] 


management_group: 5 
['user-group' 'other' 'commercial' 'parastatal' 'unknown'] 


payment: 7 
['pay annually' 'never pay' 'pay per bucket' 'unknown'
 'pay when scheme fails' 'other' 'pay monthly'] 


payment_type: 7 
['annually' 'never pay' 'per bucket' 'unknown' 'on failure' 'other'
 'monthly'] 


water_quality: 8 
['soft' 'salty' 'milky' 'unknown' 'fluoride' 'coloured' 'salty abandoned'
 'fluoride abandoned'] 


quality_group: 6 
['good' 'salty' 'milky' 'unknown' 'fluoride' 'colored'] 


quantity: 5 
['enough' 'insufficient' 'dry' 'seasonal' 'unknown'] 


quantity_group: 5 
['enough' 'insufficient' 'dry' 'seasonal' 'unknown'] 


source: 10 
['spring' 'rainwater harvesting' 'dam' 'machine dbh' 'other'
 'shallow well' 

`id`, `date_recorded`, `recorded_by` do not give usefull information to our model
<br>
Looks like `wpt_name`, `subvillage` will lead to model overfitting 
<br>
`num_private` - no description
<br>
`region_code`, `district_code`, `lga`, `ward`, `scheme_name`, `extraction_type`, `extraction_type_group`, `management`, `payment`, `water_quality`, `quantity_group`, `source`, `source_class`, `waterpoint_type` are duplicates or very simmilar

In [6]:
features = raw_df[['amount_tsh', 'funder', 'gps_height', 'installer', 'longitude', 'latitude', 
                             'basin', 'region', 'population', 'public_meeting', 'scheme_management', 
                             'permit', 'construction_year', 'extraction_type_class', 'management_group', 
                             'payment_type', 'quality_group', 'quantity', 'source_type']]

#### Import labels

In [7]:
labels = pd.read_csv('data/downloaded/train_labels.csv')
labels = labels.status_group
labels.unique()

array(['functional', 'non functional', 'functional needs repair'],
      dtype=object)

#### Concatenate features and labels

In [8]:
data = pd.concat([features,labels], axis=1)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   amount_tsh             59400 non-null  float64
 1   funder                 59400 non-null  object 
 2   gps_height             59400 non-null  int64  
 3   installer              59400 non-null  object 
 4   longitude              59400 non-null  float64
 5   latitude               59400 non-null  float64
 6   basin                  59400 non-null  object 
 7   region                 59400 non-null  object 
 8   population             59400 non-null  int64  
 9   public_meeting         59400 non-null  object 
 10  scheme_management      59400 non-null  object 
 11  permit                 59400 non-null  object 
 12  construction_year      59400 non-null  int64  
 13  extraction_type_class  59400 non-null  object 
 14  management_group       59400 non-null  object 
 15  pa

#### Convert boolean values to string

In [9]:
mask = data.applymap(type) != bool
d = {True: 'yes', False: 'no'}

data = data.where(mask, data.replace(d))

#### Save 

In [10]:
with open('data/preprocessed/data.pickle', 'wb') as file:
    pickle.dump(data, file)

#### Check number of unique values for each feature

In [11]:
for f in data:
    print(f'\n\n--------------------------------------------\n{f}: {data[f].nunique()}\n')
    print(data[f].value_counts() )



--------------------------------------------
amount_tsh: 98

0.0         41639
500.0        3102
50.0         2472
1000.0       1488
20.0         1463
            ...  
8500.0          1
6300.0          1
220.0           1
138000.0        1
12.0            1
Name: amount_tsh, Length: 98, dtype: int64


--------------------------------------------
funder: 1898

Government Of Tanzania    9084
not_filled                3635
Danida                    3114
Hesawa                    2202
Rwssp                     1374
                          ... 
Waitaliano                   1
Mahemba                      1
Nyamuhanga Maro              1
Wame Mbiki                   1
Morad                        1
Name: funder, Length: 1898, dtype: int64


--------------------------------------------
gps_height: 2428

 0       20438
-15         60
-16         55
-13         55
-20         52
         ...  
 2285        1
 2424        1
 2552        1
 2413        1
 2385        1
Name: gps_height, Lengt

#### Count missing values 

In [12]:
print('Missing values count')
for f in data:
    zero_val_len = len(data[data[f] == 0])
    if zero_val_len > 0:
        print('{0:25}{1}'.format(f,zero_val_len))

Missing values count
amount_tsh               41639
gps_height               20438
longitude                1812
population               21381
construction_year        20709


## EDA <small style="color:#000"><strong>Use visualization.ipynb file</strong></small>

## Data cleaning
Dataset will be saved 2 times with different preprocessing:
    1. Data with combined categorical data for `funder` and `installer`
    2. The same as 1, but with removed outliers, and dropped all missing values

`amount_tsh` - 41639 of 59400 values missing: better to remove whole column

In [13]:
data.drop('amount_tsh', axis=1, inplace=True)

#### Split dataset first to avoid data leakage

In [14]:
train, test = train_test_split(data,random_state=42)

#### 1. Features with combined categorical data
Decrease unique values of selected columns to 100 (set value 'Other' for less frequent columns)

In [15]:
train_funder_col_names, assigned_val, train.funder = hlp.reduce_cat_num(train.funder,'Other')
test.funder = hlp.apply_reduced_cats(test.funder,train_funder_col_names,assigned_val)

train_installer_col_names, assigned_val, train.installer = hlp.reduce_cat_num(train.installer,'Other')
test.installer = hlp.apply_reduced_cats(test.installer,train_installer_col_names,assigned_val)

#### Save

In [16]:
with open('data/preprocessed/train_comb_cats.pickle', 'wb') as file:
    pickle.dump(train, file)
    
with open('data/preprocessed/test_comb_cats.pickle', 'wb') as file:
    pickle.dump(test, file)

#### 2. Features with combined categorical data, removed outliers, and dropped all missing values

In [17]:
# train dropped missing values
train_dmv = train.copy()
train_dmv = train_dmv[train_dmv['gps_height'] != 0]
train_dmv = train_dmv[train_dmv['longitude'] != 0]
train_dmv = train_dmv[train_dmv['population'] != 0]
train_dmv = train_dmv[train_dmv['construction_year'] != 0]

hlp.print_reduce_perc(train,train_dmv)

'Dataframe length -  before: 44550, after: 27514. Size reduction: 38.24%'

In [18]:
# Check for -2.000000e-08 value
train_dmv.latitude.value_counts()

-6.994704    2
-6.975594    2
-6.978263    2
-7.102004    2
-6.962475    2
            ..
-6.518852    1
-7.738807    1
-9.725791    1
-4.429275    1
-7.159651    1
Name: latitude, Length: 27485, dtype: int64

#### Remove outliers

In [19]:
possible_outliers_cols = ['gps_height','longitude','latitude','population']
train_dmv_cleaned = train_dmv[(np.abs(stats.zscore(train_dmv[possible_outliers_cols])) < 3).all(axis=1)]

hlp.print_reduce_perc(train_dmv,train_dmv_cleaned)

'Dataframe length -  before: 27514, after: 27082. Size reduction: 1.57%'

#### Save

In [20]:
with open('data/preprocessed/train_comb_cats_cleaned.pickle', 'wb') as file:
    pickle.dump(train_dmv_cleaned, file)

## Data for submission

#### Load test dataset

In [21]:
test_df = pd.read_csv('data/downloaded/test_data.csv')
test_df.head()

id  amount_tsh date_recorded                  funder  gps_height  \
0  50785         0.0    2013-02-04                    Dmdd        1996   
1  51630         0.0    2013-02-04  Government Of Tanzania        1569   
2  17168         0.0    2013-02-01                     NaN        1567   
3  45559         0.0    2013-01-22              Finn Water         267   
4  49871       500.0    2013-03-27                  Bruder        1260   

    installer  longitude   latitude                 wpt_name  num_private  \
0        DMDD  35.290799  -4.059696  Dinamu Secondary School            0   
1         DWE  36.656709  -3.309214                  Kimnyak            0   
2         NaN  34.767863  -5.004344           Puma Secondary            0   
3  FINN WATER  38.058046  -9.418672           Kwa Mzee Pange            0   
4      BRUDER  35.006123 -10.950412          Kwa Mzee Turuka            0   

   ... payment_type water_quality quality_group      quantity  quantity_group  \
0  ...    never pay          soft          good      seasonal        seasonal   
1  ...    never pay          soft          good  insufficient    insufficient   
2  ...    never pay          soft          good  insufficient    insufficient   
3  ...      unknown          soft          good           dry             dry   
4  ...      monthly          soft          good        enough          enough   

                 source           source_type  source_class  \
0  rainwater harvesting  rainwater harvesting       surface   
1                spring                spring   groundwater   
2  rainwater harvesting  rainwater harvesting       surface   
3          shallow well          shallow well   groundwater   
4                spring                spring   groundwater   

      waterpoint_type waterpoint_type_group  
0               other                 other  
1  communal standpipe    communal standpipe  
2               other                 other  
3               other                 other  
4  communal standpipe    communal standpipe  

[5 rows x 40 columns]

#### Fill na values in categorical features

In [27]:
test_df_na_cols = test_df.columns[test_df.isna().any()].tolist()
for col in test_df_na_cols:
    test_df[col] = test_df[col].fillna('not_filled')

In [28]:
test_features = test_df[['funder', 'gps_height', 'installer', 'longitude', 'latitude', 
                             'basin', 'region', 'population', 'public_meeting', 'scheme_management', 
                             'permit', 'construction_year', 'extraction_type_class', 'management_group', 
                             'payment_type', 'quality_group', 'quantity', 'source_type']]

#### Convert boolean values to string

In [29]:
# mask = data.applymap(type) != bool
# d = {True: 'yes', False: 'no'}

test_features = test_features.where(mask, test_features.replace(d))

#### Combine categories in `funder` and `installer` columns

In [30]:
data_sm = data.copy() # Submission data

data_sm_funder_col_names, assigned_val, data_sm.funder = hlp.reduce_cat_num(data_sm.funder,'Other')
test_df.funder = hlp.apply_reduced_cats(test_df.funder,data_sm_funder_col_names,assigned_val)

data_sm_installer_col_names, assigned_val, data_sm.installer = hlp.reduce_cat_num(data_sm.installer,'Other')
test_df.installer = hlp.apply_reduced_cats(test_df.installer,data_sm_installer_col_names,assigned_val)

#### Save

In [31]:
with open('data/preprocessed/train_for_SM.pickle', 'wb') as file:
    pickle.dump(data_sm, file)

with open('data/preprocessed/test_for_SM.pickle', 'wb') as file:
    pickle.dump(test_df, file)